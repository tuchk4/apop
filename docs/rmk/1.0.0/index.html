<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>Home - Documentation</title>

    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="actions.html">actions</a><ul class='methods'><li data-type='method'><a href="actions.html#.clear">clear</a></li><li data-type='method'><a href="actions.html#.toCamelCase">toCamelCase</a></li><li data-type='method'><a href="actions.html#.toSnakeCase">toSnakeCase</a></li></ul></li><li><a href="shortcuts.html">shortcuts</a><ul class='methods'><li data-type='method'><a href="shortcuts.html#.clear">clear</a></li><li data-type='method'><a href="shortcuts.html#.toCamelCase">toCamelCase</a></li><li data-type='method'><a href="shortcuts.html#.toSnakeCase">toSnakeCase</a></li></ul></li><li><a href="utils.html">utils</a><ul class='methods'><li data-type='method'><a href="utils.html#.parseArray">parseArray</a></li><li data-type='method'><a href="utils.html#.parseObject">parseObject</a></li><li data-type='method'><a href="utils.html#.reduce">reduce</a></li></ul></li></ul>
</nav>

<div id="main">
    

    



    









    


    <section class="readme">
        <article><h1>RMK</h1><p>Object transformation without mutations.</p>
<pre class="prettyprint source"><code>REMAKE = Function(data): newData
Formula = Function(REMAKE_1 ... REMAKE_n): Function(data): newData</code></pre><p>Pseudo example:</p>
<pre class="prettyprint source lang-js"><code>import rmk from 'rmk';
import toCamelCase from '@rmk/to-camel-case';
import clear from '@rmk/clear';
import update from '@rmk/update';

const formula = rmk(
  toCamelCase(),
  clear(),
  update({
    id: data => data.id + 10
  })
);

const object = [{
  id: 1,
  groupId: 1,
  name: null
}, {
  id: 2,
  groupId: 2,
  name: ''
}];

const newObject = rmk(object);
expect(newObject).toEqual([{
  id: 10,
  group_id: 1
}, {
  id: 20,
  group_id: 2
}])</code></pre></article>
    </section>






</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.4.3</a> on Mon Feb 20 2017 17:16:35 GMT+0200 (EET) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>
</body>
</html>