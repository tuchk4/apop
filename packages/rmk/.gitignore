.DS_Store
/.git
/.idea
/npm-debug.log
/node_modules
/dist
recursive
*.json
/*.js
/*.map
!__tests__
!src
!rollup-build.js
