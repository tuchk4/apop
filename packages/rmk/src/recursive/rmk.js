import rmkActions from './../actions';
import rmkParse from './../utils/parse';

const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default recursive;
