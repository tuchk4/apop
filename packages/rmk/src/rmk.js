import rmkActions from './actions';
import rmkParse from './utils/parse';

const singleActions = {};
const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  singleActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: false,
    });

  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const rmk = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: false }),
  singleActions
);
const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default Object.assign(rmk, { recursive });
