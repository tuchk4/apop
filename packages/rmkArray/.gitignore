.DS_Store
/.git
/.idea
/npm-debug.log
/node_modules
/dist
*.json

!__tests__
!src
!rollup-build.js
/*.js
/*.map