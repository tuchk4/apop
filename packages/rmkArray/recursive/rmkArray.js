(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global['recursive/rmkArray'] = factory());
}(this, (function () { 'use strict';

/** @namespace rmkArray */

// import clear from './clear';
// import toSnakeCase from './toSnakeCase';
// import toCamelCase from './toCamelCase';
// import update from './update';
// import rename from './rename';
// import each from './each';

var rmkArrayActions = {};

var parseEntity = function parseEntity(entity, deep, config) {
  // isArray
  if (entity && entity.constructor === Array && (config.isRecursive || !config.isRecursive && deep == 0)) {
    // copy array
    entity = entity.slice(0);

    for (var j = 0, maxActions = config.actions.length; j < maxActions; j++) {
      //parse by action. copy object in utils eachKeys
      entity = config.actions[j](entity);
    }

    // deep parse array in deep
    if (config.isRecursive || !config.isRecursive && deep == 0) {
      var entitySize = entity.length;
      for (var i = 0; i < entitySize; i++) {
        entity[i] = parseEntity(entity[i], deep + 1, config);
      }
    }
  }
  return entity;
};

var rmkArrayParse = function rmkArrayParse(config) {
  return function (origin) {
    return parseEntity(origin, 0, config);
  };
};

var rmkArrayRecursiveActions = {};

Object.keys(rmkArrayActions).forEach(function (action) {

  rmkArrayRecursiveActions[action] = function (config) {
    return rmkArrayParse({
      actions: [rmkArrayActions[action](config)],
      isRecursive: true
    });
  };
});

var recursive = Object.assign(function () {
  for (var _len = arguments.length, actions = Array(_len), _key = 0; _key < _len; _key++) {
    actions[_key] = arguments[_key];
  }

  return rmkArrayParse({ actions: actions, isRecursive: true });
}, rmkArrayRecursiveActions);

return recursive;

})));
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoicm1rQXJyYXkuanMiLCJzb3VyY2VzIjpbIi4uL3NyYy9hY3Rpb25zL2luZGV4LmpzIiwiLi4vc3JjL3V0aWxzL3BhcnNlLmpzIiwiLi4vc3JjL3JlY3Vyc2l2ZS9ybWtBcnJheS5qcyJdLCJzb3VyY2VzQ29udGVudCI6WyIvKiogQG5hbWVzcGFjZSBybWtBcnJheSAqL1xuXG4vLyBpbXBvcnQgY2xlYXIgZnJvbSAnLi9jbGVhcic7XG4vLyBpbXBvcnQgdG9TbmFrZUNhc2UgZnJvbSAnLi90b1NuYWtlQ2FzZSc7XG4vLyBpbXBvcnQgdG9DYW1lbENhc2UgZnJvbSAnLi90b0NhbWVsQ2FzZSc7XG4vLyBpbXBvcnQgdXBkYXRlIGZyb20gJy4vdXBkYXRlJztcbi8vIGltcG9ydCByZW5hbWUgZnJvbSAnLi9yZW5hbWUnO1xuLy8gaW1wb3J0IGVhY2ggZnJvbSAnLi9lYWNoJztcblxuZXhwb3J0IGRlZmF1bHQge307XG4iLCJjb25zdCBwYXJzZUVudGl0eSA9IChlbnRpdHksIGRlZXAsIGNvbmZpZykgPT4ge1xuICAvLyBpc0FycmF5XG4gIGlmIChcbiAgICBlbnRpdHkgJiZcbiAgICBlbnRpdHkuY29uc3RydWN0b3IgPT09IEFycmF5ICYmXG4gICAgKGNvbmZpZy5pc1JlY3Vyc2l2ZSB8fCAoIWNvbmZpZy5pc1JlY3Vyc2l2ZSAmJiBkZWVwID09IDApKVxuICApIHtcbiAgICAvLyBjb3B5IGFycmF5XG4gICAgZW50aXR5ID0gZW50aXR5LnNsaWNlKDApO1xuXG4gICAgZm9yIChsZXQgaiA9IDAsIG1heEFjdGlvbnMgPSBjb25maWcuYWN0aW9ucy5sZW5ndGg7IGogPCBtYXhBY3Rpb25zOyBqKyspIHtcbiAgICAgIC8vcGFyc2UgYnkgYWN0aW9uLiBjb3B5IG9iamVjdCBpbiB1dGlscyBlYWNoS2V5c1xuICAgICAgZW50aXR5ID0gY29uZmlnLmFjdGlvbnNbal0oZW50aXR5KTtcbiAgICB9XG5cbiAgICAvLyBkZWVwIHBhcnNlIGFycmF5IGluIGRlZXBcbiAgICBpZiAoY29uZmlnLmlzUmVjdXJzaXZlIHx8ICghY29uZmlnLmlzUmVjdXJzaXZlICYmIGRlZXAgPT0gMCkpIHtcbiAgICAgIGxldCBlbnRpdHlTaXplID0gZW50aXR5Lmxlbmd0aDtcbiAgICAgIGZvciAobGV0IGkgPSAwOyBpIDwgZW50aXR5U2l6ZTsgaSsrKSB7XG4gICAgICAgIGVudGl0eVtpXSA9IHBhcnNlRW50aXR5KGVudGl0eVtpXSwgZGVlcCArIDEsIGNvbmZpZyk7XG4gICAgICB9XG4gICAgfVxuICB9XG4gIHJldHVybiBlbnRpdHk7XG59O1xuXG5jb25zdCBybWtBcnJheVBhcnNlID0gY29uZmlnID0+IG9yaWdpbiA9PiBwYXJzZUVudGl0eShvcmlnaW4sIDAsIGNvbmZpZyk7XG5cbmV4cG9ydCBkZWZhdWx0IHJta0FycmF5UGFyc2U7XG4iLCJpbXBvcnQgcm1rQXJyYXlBY3Rpb25zIGZyb20gJy4vLi4vYWN0aW9ucyc7XG5pbXBvcnQgcm1rQXJyYXlQYXJzZSBmcm9tICcuLy4uL3V0aWxzL3BhcnNlJztcblxuY29uc3Qgcm1rQXJyYXlSZWN1cnNpdmVBY3Rpb25zID0ge307XG5cbk9iamVjdC5rZXlzKHJta0FycmF5QWN0aW9ucykuZm9yRWFjaChhY3Rpb24gPT4ge1xuXG4gIHJta0FycmF5UmVjdXJzaXZlQWN0aW9uc1thY3Rpb25dID0gY29uZmlnID0+XG4gICAgcm1rQXJyYXlQYXJzZSh7XG4gICAgICBhY3Rpb25zOiBbcm1rQXJyYXlBY3Rpb25zW2FjdGlvbl0oY29uZmlnKV0sXG4gICAgICBpc1JlY3Vyc2l2ZTogdHJ1ZSxcbiAgICB9KTtcbn0pO1xuXG5jb25zdCByZWN1cnNpdmUgPSBPYmplY3QuYXNzaWduKFxuICAoLi4uYWN0aW9ucykgPT4gcm1rQXJyYXlQYXJzZSh7IGFjdGlvbnMsIGlzUmVjdXJzaXZlOiB0cnVlIH0pLFxuICBybWtBcnJheVJlY3Vyc2l2ZUFjdGlvbnNcbik7XG5cbmV4cG9ydCBkZWZhdWx0IHJlY3Vyc2l2ZTtcbiJdLCJuYW1lcyI6WyJwYXJzZUVudGl0eSIsImVudGl0eSIsImRlZXAiLCJjb25maWciLCJjb25zdHJ1Y3RvciIsIkFycmF5IiwiaXNSZWN1cnNpdmUiLCJzbGljZSIsImoiLCJtYXhBY3Rpb25zIiwiYWN0aW9ucyIsImxlbmd0aCIsImVudGl0eVNpemUiLCJpIiwicm1rQXJyYXlQYXJzZSIsIm9yaWdpbiIsInJta0FycmF5UmVjdXJzaXZlQWN0aW9ucyIsIk9iamVjdCIsImtleXMiLCJybWtBcnJheUFjdGlvbnMiLCJmb3JFYWNoIiwiYWN0aW9uIiwicmVjdXJzaXZlIiwiYXNzaWduIl0sIm1hcHBpbmdzIjoiOzs7Ozs7QUFBQTs7Ozs7Ozs7O0FBU0Esc0JBQWUsRUFBZjs7QUNUQSxJQUFNQSxjQUFjLFNBQWRBLFdBQWMsQ0FBQ0MsTUFBRCxFQUFTQyxJQUFULEVBQWVDLE1BQWYsRUFBMEI7O01BRzFDRixVQUNBQSxPQUFPRyxXQUFQLEtBQXVCQyxLQUR2QixLQUVDRixPQUFPRyxXQUFQLElBQXVCLENBQUNILE9BQU9HLFdBQVIsSUFBdUJKLFFBQVEsQ0FGdkQsQ0FERixFQUlFOzthQUVTRCxPQUFPTSxLQUFQLENBQWEsQ0FBYixDQUFUOztTQUVLLElBQUlDLElBQUksQ0FBUixFQUFXQyxhQUFhTixPQUFPTyxPQUFQLENBQWVDLE1BQTVDLEVBQW9ESCxJQUFJQyxVQUF4RCxFQUFvRUQsR0FBcEUsRUFBeUU7O2VBRTlETCxPQUFPTyxPQUFQLENBQWVGLENBQWYsRUFBa0JQLE1BQWxCLENBQVQ7Ozs7UUFJRUUsT0FBT0csV0FBUCxJQUF1QixDQUFDSCxPQUFPRyxXQUFSLElBQXVCSixRQUFRLENBQTFELEVBQThEO1VBQ3hEVSxhQUFhWCxPQUFPVSxNQUF4QjtXQUNLLElBQUlFLElBQUksQ0FBYixFQUFnQkEsSUFBSUQsVUFBcEIsRUFBZ0NDLEdBQWhDLEVBQXFDO2VBQzVCQSxDQUFQLElBQVliLFlBQVlDLE9BQU9ZLENBQVAsQ0FBWixFQUF1QlgsT0FBTyxDQUE5QixFQUFpQ0MsTUFBakMsQ0FBWjs7OztTQUlDRixNQUFQO0NBdkJGOztBQTBCQSxJQUFNYSxnQkFBZ0IsU0FBaEJBLGFBQWdCO1NBQVU7V0FBVWQsWUFBWWUsTUFBWixFQUFvQixDQUFwQixFQUF1QlosTUFBdkIsQ0FBVjtHQUFWO0NBQXRCOztBQ3ZCQSxJQUFNYSwyQkFBMkIsRUFBakM7O0FBRUFDLE9BQU9DLElBQVAsQ0FBWUMsZUFBWixFQUE2QkMsT0FBN0IsQ0FBcUMsa0JBQVU7OzJCQUVwQkMsTUFBekIsSUFBbUM7V0FDakNQLGNBQWM7ZUFDSCxDQUFDSyxnQkFBZ0JFLE1BQWhCLEVBQXdCbEIsTUFBeEIsQ0FBRCxDQURHO21CQUVDO0tBRmYsQ0FEaUM7R0FBbkM7Q0FGRjs7QUFTQSxJQUFNbUIsWUFBWUwsT0FBT00sTUFBUCxDQUNoQjtvQ0FBSWIsT0FBSjtXQUFBOzs7U0FBZ0JJLGNBQWMsRUFBRUosZ0JBQUYsRUFBV0osYUFBYSxJQUF4QixFQUFkLENBQWhCO0NBRGdCLEVBRWhCVSx3QkFGZ0IsQ0FBbEI7Ozs7Ozs7OyJ9
