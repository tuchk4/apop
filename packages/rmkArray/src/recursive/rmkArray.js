import rmkArrayActions from './../actions';
import rmkArrayParse from './../utils/parse';

const rmkArrayRecursiveActions = {};

Object.keys(rmkArrayActions).forEach(action => {

  rmkArrayRecursiveActions[action] = config =>
    rmkArrayParse({
      actions: [rmkArrayActions[action](config)],
      isRecursive: true,
    });
});

const recursive = Object.assign(
  (...actions) => rmkArrayParse({ actions, isRecursive: true }),
  rmkArrayRecursiveActions
);

export default recursive;
