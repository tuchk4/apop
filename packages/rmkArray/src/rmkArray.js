import filter from './actions/filter';
import join from './actions/join';
import remove from './actions/remove';
import removeElements from './actions/removeElements';
import swap from './actions/swap';
import sort from './actions/sort';
import map from './actions/map';

import parse from './utils/parse';

const sourceActions = {filter, join, remove, removeElements,  swap, sort, map};
const singleActions = {};
const recursiveActions = {};

Object.keys(sourceActions).forEach(action => {
  singleActions[action] = (...args) =>
    parse({
      actions: [sourceActions[action](...args)],
      isRecursive: false,
    });

  recursiveActions[action] = (...args) =>
    parse({
      actions: [sourceActions[action](...args)],
      isRecursive: true,
    });
});

const rmkArray = Object.assign(
  (...actions) => parse({ actions, isRecursive: false }),
  singleActions
);

const recursive = Object.assign(
  (...actions) => parse({ actions, isRecursive: true }),
  recursiveActions
);

export default Object.assign(rmkArray, { recursive: recursive });
