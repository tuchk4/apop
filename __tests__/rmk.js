import rmk from '../packages/rmk/src/rmk';
import rmkArray from '../packages/rmkArray/src/rmkArray';


var beforeArray = [];
for(let i = 15; i < 30; i++){
  beforeArray.push({
    index: i,
    date: `${new Date(Date.UTC(2001, 11, i))}`
  });
}
for(let i = 1; i <= 12; i++){
  beforeArray.push({
    index: i,
    date: `${new Date(Date.UTC(2001, i, 3))}`
  });
}

it('First Global Test', () => {
  const objectFormula = rmk(
    rmk.rename({'index': 'id'}),
    rmk.update({
      sort: (state) => (state.id * 3) + (state.id * 2) + state.id,
      case_detect: true
    }),
    rmk.recursive.toCamelCase()
  );
  const arrayFormula = rmkArray(
    rmkArray.swap(3, 4),
    rmkArray.swap(5, 6),
    rmkArray.swap(4, 5),
    rmkArray.removeElements(beforeArray[0], beforeArray[1]),
    rmkArray.filter(item => item.index % 2),
    rmkArray.remove(4,5,6,7),
    rmkArray.map(rmk.rename({'date':'birth_date'})),
    rmkArray.map(objectFormula),
    rmkArray.sort('sort', 'desc'),
  );

  const result = arrayFormula(beforeArray);

  expect(result).toMatchSnapshot();
});