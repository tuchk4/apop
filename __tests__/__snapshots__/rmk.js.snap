// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First Global Test 1`] = `
Array [
  Object {
    "birthDate": "Sun Dec 23 2001 02:00:00 GMT+0200 (EET)",
    "caseDetect": true,
    "id": 23,
    "sort": 138,
  },
  Object {
    "birthDate": "Fri Dec 21 2001 02:00:00 GMT+0200 (EET)",
    "caseDetect": true,
    "id": 21,
    "sort": 126,
  },
  Object {
    "birthDate": "Wed Dec 19 2001 02:00:00 GMT+0200 (EET)",
    "caseDetect": true,
    "id": 19,
    "sort": 114,
  },
  Object {
    "birthDate": "Mon Dec 17 2001 02:00:00 GMT+0200 (EET)",
    "caseDetect": true,
    "id": 17,
    "sort": 102,
  },
  Object {
    "birthDate": "Mon Dec 03 2001 02:00:00 GMT+0200 (EET)",
    "caseDetect": true,
    "id": 11,
    "sort": 66,
  },
  Object {
    "birthDate": "Wed Oct 03 2001 03:00:00 GMT+0300 (EEST)",
    "caseDetect": true,
    "id": 9,
    "sort": 54,
  },
  Object {
    "birthDate": "Fri Aug 03 2001 03:00:00 GMT+0300 (EEST)",
    "caseDetect": true,
    "id": 7,
    "sort": 42,
  },
  Object {
    "birthDate": "Sun Jun 03 2001 03:00:00 GMT+0300 (EEST)",
    "caseDetect": true,
    "id": 5,
    "sort": 30,
  },
  Object {
    "birthDate": "Tue Apr 03 2001 03:00:00 GMT+0300 (EEST)",
    "caseDetect": true,
    "id": 3,
    "sort": 18,
  },
]
`;
